adding upgrade code for upgrading from 0.9.7 to 0.9.8.

albertshau opened the pull request:
Will keep a set of codecs for serializing/deserializing current
objects, and a separate set for converting between versions.
Switched to use guice injection to create the codec everywhere.

albertshau committed:
adding upgrade code for upgrading from 0.9.7 to 0.9.8.
Will keep a set of codecs for serializing/deserializing current
objects, and a separate set for converting between versions.
Switched to use guice injection to create the codec everywhere.

wolf31o2 commented:
I think `loom-upgrade-tables-mysql.sql` should have versions in it. We're going to end up with a few of these, over time. Also, I think that we should do the same with the `loom-create-tables-mysql.sql` file. Either that, or we should merge them and make a single SQL that can create _or_ upgrade.

albertshau committed:
update changelog with upgrade script.

albertshau commented:
I can see why we'd want multiple upgrade files, but what would be the reason for multiple create files?  I don't think somebody would ever have to run more than one create file and if we want it for historical purposes they are always in git.

poornachandra commented on server/src/main/java/com/continuuity/loom/store/DBQueryHelper.java:
```diff
@@ -26,19 +27,30 @@
+        // and if the table already exists, we return false.  But if there was some other exception we'll throw that.
         if (!e.getSQLState().equals("X0Y32")) {
-          throw Throwables.propagate(e);
+          throw new SQLException(e);
```
You can throw the same object, right? Do you need to create a new wrapped exception?

albertshau replied:
nope you're right.

albertshau committed:
Merge remote-tracking branch 'origin/develop' into feature/db-migration

Conflicts:
	server/src/main/java/com/continuuity/loom/http/handler/LoomClusterHandler.java
	server/src/main/java/com/continuuity/loom/scheduler/CallbackScheduler.java
	server/src/main/java/com/continuuity/loom/scheduler/ClusterScheduler.java
	server/src/main/java/com/continuuity/loom/scheduler/JobScheduler.java
	server/src/main/java/com/continuuity/loom/scheduler/SolverScheduler.java
	server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java
	server/src/main/java/com/continuuity/loom/scheduler/task/TaskService.java
	server/src/main/java/com/continuuity/loom/store/cluster/SQLClusterStoreService.java
	server/src/test/java/com/continuuity/loom/http/LoomClusterHandlerTest.java
	server/src/test/java/com/continuuity/loom/scheduler/ClusterCleanupTest.java
	server/src/test/java/com/continuuity/loom/scheduler/SchedulerTest.java
	server/src/test/java/com/continuuity/loom/scheduler/SolverSchedulerTest.java

poornachandra commented on server/src/main/java/com/continuuity/loom/http/handler/LoomAdminHandler.java:
```diff
@@ -75,13 +74,13 @@
 
   @Inject
-  private LoomAdminHandler(TenantStore tenantStore, EntityStoreService entityStoreService) {
+  private LoomAdminHandler(TenantStore tenantStore, EntityStoreService entityStoreService, JsonSerde codec) {
```
Should we replace all JsonSerde injection directly with Gson injection, since we use JsonSerde only to get a Gson object?

albertshau replied:
aside from getting the gson, JsonSerde is only used to transform strings into bytes and vice versa, so we could just get rid of it and bind Gson in the codec module.  will do that.

albertshau committed:
pull request feedback.  Removing JsonSerde since its just used to
get its underlying Gson, so just binding Gson directly. Refactored
a bit so that database queries are performed by a separated
DBQueryExecutor class that gets its gson injected so that we are
not passing gson around in every method call.

Also adding version to the upgrade tables script.

albertshau committed:
adding renamed file.

poornachandra commented on server/src/main/java/com/continuuity/loom/tools/UpgradeTool.java:
```diff
@@ -0,0 +1,93 @@
+    DBConnectionPool dbConnectionPool = injector.getInstance(DBConnectionPool.class);
+    alterTables(dbConnectionPool);
+    migrateClusters(clusterStoreService);
+  }
```
We should stop the clusterStoreService at the end.

poornachandra commented:
One minor comment, you can merge after that :+1:

albertshau committed:
stop service before exiting.
